### ISSUE TITLE
[question] using self-written (not conan-center) conan recipes during a build?

### ISSUE THREAD

**Issue by justinabrahms** [conan-io/conan#12663]
We have a group internally who is trying to build open source C++ code using conan. 

We rely on some C++ code which doesn’t yet exist in the conan ecosystem (example: https://github.com/spdk/spdk ). The team tells me that there’s a pretty stringent process to get things listed (needing it to succesfully compile for multiple architectures, etc), such that it’s unlikely they’ll be able to submit a 3rd party recipe to conan for these dependencies. Additionally, it’s not likely that the upstreams will take the conanfile contributions. 

Is there a way to have 3rd party recipes within conan such that we can build this out in the open?

Our alternative it to use an internally hosted conan instance and do all builds internally.. but doing that behind closed doors has several questions and concerns associated that we’d love to avoid.

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).

**Comment by memsharded** [comment]
Hi @justinabrahms 

> We rely on some C++ code which doesn’t yet exist in the conan ecosystem (example: https://github.com/spdk/spdk ). The team tells me that there’s a pretty stringent process to get things listed (needing it to succesfully compile for multiple architectures, etc), such that it’s unlikely they’ll be able to submit a 3rd party recipe to conan for these dependencies. Additionally, it’s not likely that the upstreams will take the conanfile contributions.

Yes, it is true that the process to contribute to ConanCenter has some requirements, like working in multiple platforms (Windows, Linux, OSX), because the expectations from users of ConanCenter are also very high, and having a bad experience because some packages do not work well. But also the good part is that there is lot of automation, and many reviewers and contributors can help to have high quality recipes, and also to evolve them with new conan features, new dependencies, etc. As always it is a balance, and there are trade-offs.

> Is there a way to have 3rd party recipes within conan such that we can build this out in the open?

You can have public github repos with your own recipes, or even a public fork with your own recipes that you don't contribute upstream, but you can keep in sync. This is a good practice for some production environments, in which building your own binaries is desired (many of our users have very strong policies so they cannot use binaries from the internet, so they always build their own binaries).

> Our alternative it to use an internally hosted conan instance and do all builds internally.. but doing that behind closed doors has several questions and concerns associated that we’d love to avoid.

Not necessarily. Most of our users actually have their own server instance. Even just storing a copy of the packages from ConanCenter is a recommended practice. Also the majority of our users create and maintain from a few to many hundreds of their packages. Maintaining another package for a third party library is not a big deal.

Does this help with your concerns?

**Comment by justinabrahms** [comment]
@memsharded I'm trying to build an open source thing. I don't want to maintain my own personal conan server where arbitrary artifacts exist. I want it to be inspectable by the people consuming my open source (otherwise.. they may have legitimate questions about the provenance of the build).

It sounds like if we put our recipes in a public git repo, we should be able to reference them from our conan build process without having to go through conan-center? Is that right? Can you point me to any documentation on this? I'm hoping I can just have a recipe file and not have to also host all of the actual source code content as well. (would be easier for us to maintain and keep in sync)

**Comment by memsharded** [comment]
But it also depends on many other factors:

- If you are using most packages from ConanCenter, are you ready for aggressive updates from the community, and continuous need to update to latest Conan to work correctly?
- How many packages will you use from ConanCenter, and how many you expect to maintain yourself for third-parties? If it is only 1 or 2, then contributing them to ConanCenter will probably be worth it
- What level of reproducibility do you need for older releases of your own thing? How do you plan to achieve it?

> It sounds like if we put our recipes in a public git repo, we should be able to reference them from our conan build process without having to go through conan-center? Is that right? Can you point me to any documentation on this? I'm hoping I can just have a recipe file and not have to also host all of the actual source code content as well. (would be easier for us to maintain and keep in sync)

I don't think we have explicit docs for this, I have commented something similar today in https://github.com/conan-io/conan/issues/10838#issuecomment-1337090915:

- Basically provide a ``conan-center-index`` fork with your own script that does an ``export`` of your own package recipes to put them in the cache, and then build the binaries from there with ``--build=missing``.
- You don' need to store the code. The ``conan-center-index`` repo contains just references to the URL of the original repos in their ``conandata.yml`` of the recipes.

**Comment by memsharded** [comment]
The ``local-recipes-index`` feature was introduced some time ago with exactly this goal in mind: https://blog.conan.io/2024/04/23/Introducing-local-recipes-index-remote.html

I am closing this ticket as solved by that feature, please create new tickets for any further question that you might have. Thanks!
